mod-clans/src/main/java/gg/solarmc/loader/clans/Clan.java

```diff
@@ -0,0 +1,407 @@
+     * @param receiver gg.solarmc.loader.clans.ClanDataObject to remove
+     * @return Accurate set of ClanMembers post transaction
+     * @throws IllegalStateException if member already has a clan
+     * @throws IllegalArgumentException if member tried to remove was the leader of the clan, or if they are not
+     * present in the clan
+     */
+    public Set<ClanMember> removeClanMember(Transaction transaction, ClanDataObject receiver) {
+        if (receiver.getClan(transaction).isPresent()) throw new IllegalStateException("Receiver already has clan!");
+
+        ClansClanMembershipRecord rec = transaction.getProperty(DSLContext.class)
+                .fetchOne(CLANS_CLAN_MEMBERSHIP,CLANS_CLAN_MEMBERSHIP.CLAN_ID.eq(this.clanID).and(CLANS_CLAN_MEMBERSHIP.USER_ID.eq(receiver.getUserId())));
+
+        if (rec == null) throw new IllegalArgumentException("Member is not part of clan!");
+        if (receiver.isSimilar(leader)) throw new IllegalArgumentException("Tried to remove leader of clan!");
+
+        rec.delete();
+
+        receiver.setCachedClan(null);
+
+        return getClanMembers(transaction);
+    }
+
+    /**
+     * Adds clan as an ally
+     * @param transaction the tx
+     * @param receiver the receiver
+     * @return whether record was inserted or not.
+     * @throws IllegalArgumentException if receiver is the same gg.solarmc.loader.clans.Clan as this object
+     * @throws IllegalStateException if receiver or this object already has an ally
+     */
+    public boolean addClanAsAlly(Transaction transaction, Clan receiver) {
+        if (receiver.getID() == this.clanID) throw new IllegalArgumentException("Tried to mark clan ally as same clan!");
+        if (this.getAlliedClan(transaction).isPresent()) throw new IllegalStateException("Sender already have an ally!");
+        if (receiver.getAlliedClan(transaction).isPresent()) throw new IllegalStateException("Selected clan already has an ally!");
+
+        int res = transaction.getProperty(DSLContext.class)
+                .insertInto(CLANS_CLAN_ALLIANCES)
+                .columns(CLANS_CLAN_ALLIANCES.CLAN_ID,CLANS_CLAN_ALLIANCES.ALLY_ID)
+                .values(this.clanID,receiver.getID())
+                .execute();
+
+        this.alliedClan = receiver.getID();
+        receiver.alliedClan = this.getID();
+
+        return res == 1;
+    }
+
+    /**
+     * Revokes ally. Can be called from either clan. Revokes alliance for both.
+     * @param transaction the tx
+     * @return whether operation was successful or not
+     */
+    public boolean revokeAlly(Transaction transaction) {
+
+        int i = transaction.getProperty(DSLContext.class)
+                .deleteFrom(CLANS_CLAN_ALLIANCES)
+                .where(CLANS_CLAN_ALLIANCES.CLAN_ID.eq(this.clanID))
+                .or(CLANS_CLAN_ALLIANCES.ALLY_ID.eq(this.clanID))
+                .execute();
+
+        this.getAlliedClan(transaction).ifPresent(c -> c.alliedClan = null);
```

Comment:
This call to getAlliedClan may produce another database query.

I see that what you are trying to do is uncache the allied clan. For that, perhaps we need a better caching strategy, e.g. one which is centralized in the manager and therefore easier to invalidate cache entries with.

Reply:
Detail your plans for this on discord

Reply:
Smart idea - not sure why i didn't think of it first! Should be in tomorrow's commit ( or tonight's if i'm not lazy)

Reply:
Fixed